### PR TITLE
Modify host shutdown logic from FunctionDispatcher

### LINKED ISSUE #4765: Reduce number of in-flight requests when shutdown is initiated
When application shutdown is initiated, there is a default grace period of 5 seconds  to allow webhost to complete any in flight requests after which the shutdown is enforced irrespective of whether the requests complete or not. In case of out of proc languages, any in-flight requests caught within this period but not complete will start throwing "Did not find any initialized language workers". This is because the shutdown enforcement has begun, but there aren't any language workers to process the incomplete, in-flight requests.

Update:
=> Changes were attempted to stop host from receiving any in flight requests by updating the ScriptHostState to Error but that affected only http requests, because trigger messages do not go through the middleware pipeline.

=> Since one of the causes also happens to be shutdown being initiated from a 3rd worker exit error (even if the previous 2 were hours or days behind) even if this can be resolved from just a re-initialization of the worker channels, changes can be made such that shutdown is enforced only when 3 language worker process exits happen within a very short interval of time.

### PR DESCRIPTION
**Pull request by yojagad**
Fixes #4765

### CODE REVIEW

File: src/WebJobs.Script.WebHost/Program.cs [hunk 1]
```diff
@@ -72,6 +72,7 @@ public static IWebHostBuilder CreateWebHostBuilder(string[] args = null)
                     loggingBuilder.AddWebJobsSystem<WebHostSystemLoggerProvider>();
                     loggingBuilder.AddDeferred();
                 })
+                .UseShutdownTimeout(TimeSpan.FromMilliseconds(1)) // Default value is 5 seconds (https://github.com/aspnet/AspNetCore/blob/master/src/Hosting/Hosting/src/Internal/WebHostOptions.cs#L70)
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Program.cs]: I don't think we should change this - we want there to be a grace period in general to allow logging and other operations to complete.

Instead, for language worker in flight requests, can't we just handle host shutdown more gracefully?

File: src/WebJobs.Script/Eventing/Rpc/WorkerErrorEvent.cs [hunk 2]
```diff
@@ -12,10 +12,13 @@ internal WorkerErrorEvent(string language, string workerId, Exception exception)
...
             Exception = exception;
             Language = language;
+            TimeOfCreation = DateTime.UtcNow;
         }
 
         internal string Language { get; private set; }
 
         public Exception Exception { get; private set; }
+
+        public DateTime TimeOfCreation { get; private set; }
```
Comment (pragnagopa) [inline src/WebJobs.Script/Eventing/Rpc/WorkerErrorEvent.cs]: Nit: Rename TimeOfCreation  to CreatedAt

File: src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs [hunk 3]
```diff
@@ -275,6 +275,26 @@ private async Task RestartWorkerChannel(string runtime, string workerId)
             }
         }
 
+        private void ModifyErrorBucketAndAdd(WorkerErrorEvent currentErrorEvent, TimeSpan threshold = default)
+        {
+            if (threshold == default)
+            {
+                threshold = TimeSpan.FromSeconds(10);
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs]: Process startup times could vary from few seconds to minutes. We would like ensure that language worker is healthy and is executing invocations for sometime. Threshold of 30mins sounds good. Also, add this is a constant to LanguageWorkerConstants.cs

File: src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs [hunk 4]
```diff
@@ -246,7 +246,7 @@ public async void WorkerError(WorkerErrorEvent workerError)
             if (!_disposing)
             {
                 _logger.LogDebug("Handling WorkerErrorEvent for runtime:{runtime}, workerId:{workerId}", workerError.Language, workerError.WorkerId);
-                _languageWorkerErrors.Add(workerError.Exception);
+                ModifyErrorBucketAndAdd(workerError);   // This either empties the stack and adds the workerError or just adds it to the stack depending on the interval between errors.
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs]: Rename method to: AddOrUpdateErroBucket. Delete comment and method name is self explanatory

File: src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs [hunk 5]
```diff
@@ -275,6 +275,26 @@ private async Task RestartWorkerChannel(string runtime, string workerId)
...
+            {
+                if ((currentErrorEvent.TimeOfCreation - top.TimeOfCreation) > threshold)
+                {
+                    while (!_languageWorkerErrors.IsEmpty)
+                    {
+                        _languageWorkerErrors.TryPop(out _);
+                    }
+                }
+            }
+            _languageWorkerErrors.Push(currentErrorEvent);
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs]: Add unit tests in FunctionDispatcherTests.cs

File: src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs [hunk 6]
```diff
@@ -246,7 +246,7 @@ public async void WorkerError(WorkerErrorEvent workerError)
             if (!_disposing)
             {
                 _logger.LogDebug("Handling WorkerErrorEvent for runtime:{runtime}, workerId:{workerId}", workerError.Language, workerError.WorkerId);
-                _languageWorkerErrors.Add(workerError.Exception);
+                AddOrUpdateErrorBucket(workerError);   // This either empties the stack and adds the workerError or just adds it to the stack depending on the interval between errors.
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs]: delete comment

File: src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs [hunk 7]
```diff
@@ -275,6 +275,26 @@ private async Task RestartWorkerChannel(string runtime, string workerId)
...
+                threshold = TimeSpan.FromMinutes(LanguageWorkerConstants.WorkerRestartErrorIntervalThresholdInMinutes);
+            }
+
+            if (_languageWorkerErrors.TryPeek(out WorkerErrorEvent top))
+            {
+                if ((currentErrorEvent.CreatedAt - top.CreatedAt) > threshold)
+                {
+                    while (!_languageWorkerErrors.IsEmpty)
+                    {
+                        _languageWorkerErrors.TryPop(out _);
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs]: add logging. with details on which worker error is popping out

File: src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs [hunk 8]
```diff
@@ -275,6 +275,26 @@ private async Task RestartWorkerChannel(string runtime, string workerId)
             }
         }
 
+        private void AddOrUpdateErrorBucket(WorkerErrorEvent currentErrorEvent, TimeSpan threshold = default)
+        {
+            if (threshold == default)
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs]: is threshold needed as  part of arguments?

File: src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs (hunk 6 above)
Comment (yojagad) [reply src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs]: Oops, sorry. Thought I did that

File: src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs (hunk 8 above)
Comment (yojagad) [reply src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs]: Maybe not

File: src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs [hunk 9]
```diff
@@ -275,20 +275,18 @@ private async Task RestartWorkerChannel(string runtime, string workerId)
...
 
             if (_languageWorkerErrors.TryPeek(out WorkerErrorEvent top))
             {
                 if ((currentErrorEvent.CreatedAt - top.CreatedAt) > threshold)
                 {
                     while (!_languageWorkerErrors.IsEmpty)
                     {
-                        _languageWorkerErrors.TryPop(out _);
+                        _languageWorkerErrors.TryPop(out WorkerErrorEvent popped);
+                        _logger.LogInformation($"Popping out errorEvent createdAt:{popped.CreatedAt} Language:{popped.Language} Name:{popped.Name} Source:{popped.Source} Exception:{popped.Exception}");
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs]: change to LogDebug

File: src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs [hunk 10]
```diff
@@ -275,20 +275,18 @@ private async Task RestartWorkerChannel(string runtime, string workerId)
...
         }
 
-        private void AddOrUpdateErrorBucket(WorkerErrorEvent currentErrorEvent, TimeSpan threshold = default)
+        private void AddOrUpdateErrorBucket(WorkerErrorEvent currentErrorEvent)
         {
-            if (threshold == default)
-            {
-                threshold = TimeSpan.FromMinutes(LanguageWorkerConstants.WorkerRestartErrorIntervalThresholdInMinutes);
-            }
+            TimeSpan threshold = TimeSpan.FromMinutes(LanguageWorkerConstants.WorkerRestartErrorIntervalThresholdInMinutes);
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs]: Since its a constant, you can add threshold as private readonly field

**Pull request merged**